File: tree/114.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func flatten(root *TreeNode)  {
	if root == nil {
		return
	}
    head := TreeNode{-1, nil, nil}
    dummy := &head
	list := make([]*TreeNode, 0)
	list = append(list, root)
	for len(list) > 0 {
		root = list[len(list)-1]
		list = list[:len(list)-1]
        dummy.Right = root
        dummy.Left = nil
        dummy = root
		if root.Right != nil {
			list = append(list, root.Right)
		}
		if root.Left != nil {
			list = append(list, root.Left)
		}
	}
    root = head.Right
	return    
}

File: tree/114.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if (!root)
            return;
        std::stack<TreeNode*> s;
        TreeNode *pre = new TreeNode{-1}, *dummy = pre;
        s.emplace(root);
        while (!s.empty()) {
            root = s.top();
            s.pop();
            pre->right = root;
            if (root->right) {
                s.emplace(root->right);
            }
            if (root->left) {
                s.emplace(root->left);
            }
            root->left = nullptr;
            pre = pre->right;
        }
        root = dummy->right;
        return;
    }
};
